libllvm: update from 9.0.1 to 10.0.0

buttaface opened the pull request:
Took the opportunity to clean up a bit and reorganize which packages come with what, see comments below. I only built for aarch64, going to try installing locally and building with these packages now.

Disabled the build in this pull for now, until all the tweaks are reviewed and I get the lldb portion done.

buttaface commented on packages/libllvm/build.sh:
```diff
@@ -1,23 +1,19 @@
-bin/clang-offload-bundler
-bin/git-clang-format
-bin/macho-dump
+bin/clang-offload-wrapper
```
These clang binaries might be useful, while `macho-dump` isn't built anymore and the last wrapper is new and isn't in the clang package for Arch linux, so left it out.

buttaface commented on packages/libllvm/build.sh:
```diff
@@ -96,16 +89,21 @@ termux_step_pre_configure() {
 		termux_error_exit "Invalid arch: $TERMUX_ARCH"
 	fi
 	# see CMakeLists.txt and tools/clang/CMakeLists.txt
-	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+=" -DLLVM_DEFAULT_TARGET_TRIPLE=$LLVM_DEFAULT_TARGET_TRIPLE"
```
[Set to `LLVM_HOST_TRIPLE` by default](https://github.com/llvm/llvm-project/blob/d4579b7ef696dc932f3174896ee5de7d235c6462/llvm/CMakeLists.txt#L647), so need to set it separately.

its-pointless replied:
Host and target are different. They have to be explicitly set as the same. This is due to jit compiler. Last time it wasn't set like that jit compiled and tried to run x86_64 code on aarch64. It then fails...

buttaface replied:
Not sure what you mean, the host triple is also set to this just two lines below. The LLVM CMake config then sets the default target triple to be the same as the host triple on Android.

its-pointless replied:
When using jit for say julialang it first v

its-pointless replied:
Jit compiles code then runs because the default target and host can be different, say default target is i686 and the host is x86_64. This means jit will use x86_64 and because it needs to run on host. Jit code is more like a script that then compiles it first time you run it. 
For us last time the build host remained default target were different causing the jit to use x86_64. Since that was the host it was built on.

buttaface replied:
I get that, but [the host triple is set to Android just two lines below](https://github.com/termux/termux-packages/pull/5136/files#diff-fa5845060496857161950d6dca5e1ba3R93), and then LLVM sets the default target triple to the host triple in the code I linked in my original comment, so they should always be the same now too.

buttaface commented on packages/libllvm/build.sh:
```diff
@@ -119,16 +117,3 @@ termux_step_post_make_install() {
-		-e "s|@LLVM_TARGET_ARCH@|$LLVM_TARGET_ARCH|g" \
-		-e "s|@LLVM_DEFAULT_TARGET_TRIPLE@|$LLVM_DEFAULT_TARGET_TRIPLE|g" \
-		-e "s|@TERMUX_ARCH@|$TERMUX_ARCH|g" > $TERMUX_PREFIX/bin/llvm-config
-	chmod 755 $TERMUX_PREFIX/bin/llvm-config
```
This was quite weird. First off all, this script was never used, from the very beginning in #917, as you have to place it in `$TERMUX_PKG_MASSAGEDIR/$TERMUX_PREFIX`. Instead, there's already a `llvm-config` executable, at least in the LLVM 9 package, and it seems to work fine.

buttaface commented on packages/libllvm/build.sh:
```diff
@@ -119,16 +117,3 @@ termux_step_post_make_install() {
-		-e "s|@TERMUX_ARCH@|$TERMUX_ARCH|g" > $TERMUX_PREFIX/bin/llvm-config
-	chmod 755 $TERMUX_PREFIX/bin/llvm-config
-	cp $TERMUX_PKG_HOSTBUILD_DIR/bin/llvm-tblgen $TERMUX_PREFIX/bin
-	cp $TERMUX_PKG_HOSTBUILD_DIR/bin/clang-tblgen $TERMUX_PREFIX/bin
```
@its-pointless, you added these two, but as above, they're not packaged since they aren't in the massage dir (plus, these are the host binaries built for x86_64, probably a typo). I notice that an AArch64 `llvm-tblgen` is already packaged currently, but `clang-tblgen` isn't installed by the LLVM config by default: would you like me to add it?

Grimler91 replied:
When we added the fast-build `-i` option we were having problems with packages trying to execute llvm-tblgen and clang-tblgen, the reason for the host binaries being installed here is probably related to that. I think we solved it some other way do so can probably remove it now though

buttaface replied:
Oh, is that why the `llvm-config` script is there too, not for the LLVM packages but for other packages that need to call `llvm-config` on the host linux build server? I just checked and I see that the rust and lldb package build scripts call it.

buttaface replied:
>we need it for dependent packages

No, that's downloaded externally, see the two lines above the one you pasted.

Grimler91 replied:
Yes, I guess its-pointless put the needed files in a tar and uploaded them to his repo. Maybe updated version is needed for lldb to build with llvm 10

its-pointless replied:
this is probably best left until after ndk 21b is released. All the fixes for 21b are completed afaik so its probably in testing. Unless recent events have forced a reschedule at google it should be a week or 2 away.

buttaface replied:
I see no reason to wait, the NDK sysroot and clang are independent. In fact, I was surprised to find yesterday that this repo goes out of its way not to use the NDK sysroot package when cross-compiling Termux packages, unlike for on-device builds.

its-pointless replied:
More about bandwidth managemen5

its-pointless replied:
More about bandwidth managemen5

its-pointless replied:
it becomes a more significant issue since rust will also need to be updated.

buttaface replied:
OK, I don't know how big a deal updating Rust is, I don't use it.

buttaface replied:
Tried updating the LLVM that the Rust package links against, let's see if it works.

buttaface commented on packages/libllvm/clang.subpackage.sh:
```diff
@@ -5,11 +5,17 @@ bin/*clang*
+include/omp.h
 lib/clang
+lib/cmake/clang
+lib/libclang*so
```
The `scan-build` and `scan-view` binaries are normally packaged with clang and `git-clang-format` could be useful. The `omp.h` header should be included here, since the library is too.

The `libclang*so` shared libraries were previously included with the `libllvm` package, and there are still 34 `libclang*a` static libraries included with `libllvm-static`. I can keep them there, include them here, or create a separate `libclang-static`, whatever is wanted.

buttaface commented on packages/libllvm/build.sh:
```diff
@@ -116,19 +114,21 @@ termux_step_post_make_install() {
-		-e "s|@TERMUX_PKG_SRCDIR@|$TERMUX_PKG_SRCDIR|g" \
-		-e "s|@LLVM_TARGET_ARCH@|$LLVM_TARGET_ARCH|g" \
 		-e "s|@LLVM_DEFAULT_TARGET_TRIPLE@|$LLVM_DEFAULT_TARGET_TRIPLE|g" \
-		-e "s|@TERMUX_ARCH@|$TERMUX_ARCH|g" > $TERMUX_PREFIX/bin/llvm-config
```
Restored and updated the llvm-config script and noticed that these two `*_ARCH` variables are unused now.

buttaface commented on packages/libllvm/llvm-config.in:
```diff
@@ -40,68 +40,80 @@ Typical components:
 prefix=@TERMUX_PREFIX@
 has_rtti=NO
 CPPFLAGS="-I${prefix}/include -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
-CFLAGS="${CPPFLAGS} ${CFLAGS}  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
```
I guess the second `CFLAGS` is to get those flags if they're set in the environment? It works for that, but I checked and the `llvm-config` binary doesn't do that. No idea why the `*_MACROS` flags were repeated here, considering they were already added by the preprended `CPPFLAGS`.

buttaface commented on packages/libllvm/llvm-config.in:
```diff
@@ -40,68 +40,80 @@ Typical components:
-CFLAGS="${CPPFLAGS} ${CFLAGS}  -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
-CXXFLAGS="${CFLAGS} -fvisibility-inlines-hidden -Wcast-qual -Wnon-virtual-dtor -std=c++11 -fno-exceptions"
+CFLAGS="${CPPFLAGS} ${CFLAGS}"
+CXXFLAGS="${CFLAGS} -std=c++14 -fno-exceptions"
```
The llvm-config binary doesn't have those three flags I removed, let me know if any should be kept.

buttaface commented:
I just used the LLVM packages from this pull to build the entire Swift 5.2.1 package on-device without a problem, looks good. I'll get lldb built next and this should be ready to go. In the meantime, I'd appreciate any review on this libllvm package.

buttaface commented on packages/lldb/build.sh:
```diff
@@ -1,39 +1,45 @@
 -DLLVM_DIR=$TERMUX_PREFIX/lib/cmake/llvm
--DClang_DIR=$TERMUX_PREFIX/lib/cmake/clang
--DLLVM_NATIVE_BUILD=$TERMUX_PREFIX/bin
+-DLLVM_TABLEGEN=$TERMUX_PKG_HOSTBUILD_DIR/llvm/bin/llvm-tblgen
```
The first three `DISABLE_` flags are ignored just like `LLVM_CONFIG`, it figures out `Clang_DIR` based on the `LLVM_DIR`, and specifying `LLVM_TABLEGEN` replaces `LLVM_NATIVE_BUILD`.

buttaface commented on packages/lldb/build.sh:
```diff
@@ -1,39 +1,45 @@
+
+	cd ..
+	cmake -G Ninja $TERMUX_PKG_SRCDIR
+	ninja -j $TERMUX_MAKE_PROCESSES lldb-tblgen
```
Rather than pre-building and downloading these, just build them from scratch.

buttaface commented on packages/lldb/build.sh:
```diff
@@ -1,39 +1,45 @@
+}
+
+termux_step_make() {
+	ninja -w dupbuild=warn -j $TERMUX_MAKE_PROCESSES all docs-lldb-man
```
For some reason, it doesn't build the man page by default.

buttaface commented on packages/lldb/utils-TableGen-LLDBTableGen.cpp.patch.beforehostbuild:
```diff
@@ -0,0 +1,14 @@
+                           "Generate lldb property enum definitions")));
+ 
++static cl::opt<bool>
++WriteIfChanged("write-if-changed", cl::desc("Only write output if it changed, does nothing"));
```
The LLVM CMake config tries to pass this flag to lldb-tablegen, but it doesn't have it, so I just copied this from llvm-tablegen and had it do nothing.

buttaface replied:
The reason it didn't have this flag is because the host lldb-tblgen was linked against the system LLVM 9 then, whereas I'm now linking against LLVM 10 built in the previous step so it works and this patch wasn't needed anymore.

buttaface commented:
Got lldb built for AArch64 and seems to be working for me locally. I tried bumping the Rust package to use this, let's see if it works.

buttaface commented:
While lldb built fine for me on my Arch linux build server using this pull, it wouldn't build the host lldb-tblgen on this Ubuntu CI, probably because it was linking against the host LLVM 9, so I added another commit to link against the just-built libraries from LLVM 10 that accompany llvm-tblgen, which made that earlier flag I added unnecessary. However, that caused problems because the lldb standalone config looks for clang, even though that isn't needed for lldb-tblgen, so I just patched those clang references out for the host build.

libllvm and lldb are building fine now, all that remains is rust.

buttaface commented:
I don't know what to make of the rust error, something about `llvm-config --version` not working (which I'm not able to reproduce), everything else builds fine.

buttaface commented:
Looks like it was finally working, it just ran out of space somewhere in the Rust build. I removed the Rust commit: what we could try is merging this llvm/lldb pull alone once it passes the CI and then I'll submit the same Rust commit in a separate pull, so it can finish building this time.

Once the Rust package is built, you can verify that the Rust package works and either merge, or revert this LLVM 10 pull till later. Let me know.

buttaface commented:
Thought of another way to test Rust, now that both the llvm and lldb updates built fine. I simply added two more commits, one to revert the lldb commit, another to update Rust. I don't think the CI will build lldb anymore and hopefully it'll clear up enough disk space for the Rust build to finish. If those Rust packages built by the CI work fine, then the commit reverting lldb can be removed and everything is ready to go.

buttaface commented:
Heh, the LLVM/Rust combo ran out of disk space on the CI too.

peter9811 commented:
It's possible get a update to version 11?

Grimler91 commented on packages/libllvm/build.sh:
```diff
@@ -96,16 +89,17 @@ termux_step_pre_configure() {
 	else
-		cp ../src/projects/openmp/runtime/exports/common.min.ompt.optional/include/omp.h $TERMUX_PREFIX/include
-	fi	
+		cp $TERMUX_PKG_SRCDIR/projects/openmp/runtime/exports/common.min.ompt.optional/include/omp.h $TERMUX_PREFIX/include
```
I am getting an error here when I try to build locally for aarch64. omp.h is at 
`/home/builder/.termux-build/libllvm/src/projects/openmp/runtime/exports/common.ompt.optional/include/omp.h` 
instead of 
`/home/builder/.termux-build/libllvm/src/projects/openmp/runtime/exports/common.min.ompt.optional/include/omp.h`
Seems to work for the CI though, weird

buttaface replied:
I changed nothing for this other than adding the prefix `$TERMUX_PKG_SRCDIR` to use the absolute path. I only built this package for AArch64 on Arch linux x86_64, didn't build for ARM myself to make sure that separate path was right.

its-pointless replied:
Yeah i have compiled with both. On old docker image it worked, on new it doesn't. So its nobodies fault.

its-pointless commented:
Updated docker image might be the cause

Grimler91 commented:
@buttaface could you rebase against master, and fix the `common.min.ompt.optional->common.ompt.optional` since it seems to be required for new docker image?

its-pointless commented:
I have already merged his libllvm and the fix into my r21b pull

buttaface commented:
Sounds good, you can just merge it there.

Grimler91 commented:
@its-pointless sorry, didn't see that, great!
